### ISSUE TITLE
MoE cache size >= 16 produces degenerate Qwen3.8 output with bounded host-pinned budget

### ISSUE THREAD

**Issue by GenerelSchwerz** [GenerelSchwerz/llama.cpp#99]
## Update: bounded host-pinned staging is the trigger

> [!IMPORTANT]
> Follow-up testing isolates the failure to the **positive bounded host-pinned budget / staged-source path**, rather than to the cache slot count by itself.

- `--moe-expert-cache-host-pinned-mb 0` restores coherent generation at the affected cache sizes.
- In this fork, `0` means **no bounded cap: attempt to pin/register the complete eligible host source set**. It does not disable host pinning.
- With the positive bounded budget from the original report (`22888` MiB), cache size 8 produces coherent output, while cache size 16 and larger produces degenerate repetition.
- Cache size 8 likely avoids the broken grouped staged path and runs legacy/cached `mul_mat_id` decode. That explanation remains an inference until a matching run confirms the final grouped counters and per-owner telemetry.

### Revised diagnosis

The cache-size threshold selects the execution path, but bounded pinning is the known trigger. Once cache size 16 makes grouped decode eligible, sources outside the positive pin budget use staged/pageable materialization and can silently produce incorrect expert data. Full pinning via `--moe-expert-cache-host-pinned-mb 0` avoids that path and resolves the observed degeneration.

This remains a correctness bug: a positive bounded pin budget must either produce the same valid expert data as full pinning or reject the configuration explicitly. It must not continue with corrupted logits and zero error counters.

### Current workaround

Use:

```text
--moe-expert-cache-host-pinned-mb 0
```

when the system can pin/register the required host memory. On systems that require a positive bounded budget, keep the cache below the grouped-decode eligibility threshold or disable the affected grouped cache path until bounded staging is fixed. Cache size 8 is coherent in the reported setup, but should not be treated as proof that bounded grouped staging is correct.

---

## Summary

Qwen3.8-Flash-Next produces degenerate repeated output when the MoE expert cache has 16 or more slots under a bounded host-pinned budget. The same prompt is coherent with the cache disabled or set to 8 slots.

The server reports no grouped-execution error counters. This appears to be silent expert-data corruption in the staged path rather than a reported fallback failure.

## Environment

- Fork commit: `907a73da9a149faa8c42ccde890f1d575586810f`
- Branch: `moe-cache`
- OS: Windows 11 build 26200
- GPU: NVIDIA RTX 4090 24 GB
- CPU: AMD Ryzen 9 5950X
- RAM: 128 GB
- CUDA: 13.4.92
- Model: `bartowski/Qwen3.8-Flash-Next-GGUF`, Q4_K_S, three shards, about 113 GB
- Architecture: `qwen4exp`

## Command

```text
llama-server.exe \
  --model ...Q4_K_S-00001-of-00003.gguf \
  --host 127.0.0.1 \
  --port 8080 \
  -ngl all \
  -fit off \
  -c 65536 \
  -b 512 \
  -ub 512 \
  -np 1 \
  -t 16 \
  -fa on \
  -ctk q8_0 \
  -ctv q8_0 \
  -kvo \
  --load-mode none \
  --moe-expert-cache-size 16 \
  --moe-expert-cache-host-pinned-mb 22888 \
  --cache-ram 0 \
  --jinja \
  --no-warmup \
  --no-reasoning-preserve \
  --experimental-logs
```

The test used one fresh conversation and the prompt `write a hello world function in C++`.

## Cache-size threshold

Only `--moe-expert-cache-size` was changed:

| Slots | Result | Decode speed |
| ---: | --- | ---: |
| 0 | Coherent | Not recorded |
| 8 | Coherent | About 4-5 tokens/s |
| 16 | Degenerate repetition such as `The///////...` | About 41-43 tokens/s in the attached run |
| 24 | Degenerate repetition | Not recorded |
| 56 | Degenerate repetition | About 40-65 tokens/s |

Sampling changes did not affect the result. `--lazy-mode on` with `--load-mode none` also did not affect it.

## Host-budget behavior

At load time, 33 consecutive groups, 15 through 47, are rejected for direct host registration:

```text
moe-cache-host: admission group=15 ... reason=source_budget_exhausted cuda=none
...
moe-cache-host: admission group=47 ... reason=source_budget_exhausted cuda=none
```

Each rejected group has three banks and about 1.52-1.57 GB of source data. These groups should use staged transport under the configured host budget.

A clean size-56 shutdown reported:

```text
admitted_banks=32112
fallback=0
rollback=0
required_unsupported=0
prepare_error=0
finish_error=0
```

The full grouped telemetry line, including `host_calls`, `host_ops`, `host_split_ops`, `strategy_switches`, `h2d_banks`, and `h2d_bytes`, can be supplied if needed.

## Observed behavior

- Prompt processing completes normally.
- Decode runs at high throughput, but output is degenerate repetition instead of coherent text.
- GPU utilization is 96-98 percent.
- VRAM use is only about 9.1 GB of 24 GB.
- RAM use is about 73 GB.
- No grouped error counters identify the bad output.

The low VRAM use and direct-registration warnings indicate that the host-pinned budget, rather than VRAM capacity, selects the affected path.

## Shutdown behavior

With 16 slots, the server repeatedly remained at `cleaning up before exit...` for several minutes after Ctrl+C. A second Ctrl+C force-terminated it before the final summary in four out of four attempts. It is not yet clear whether this is a deadlock or slow teardown of staged resources.

## Expected behavior

- Changing the cache slot count must not change generated text from coherent output to degenerate repetition.
- Groups rejected from direct registration must assemble the correct expert data through staged transport.
- If a staged configuration is unsupported, the server should fail explicitly rather than continue with incorrect logits.
- Shutdown should complete without requiring a second interrupt.

## Relevant log excerpt

```text
0.00.043.431 I srv load_model: loading model 'D:\models\qwen3.8-flash-next...\Qwen3.8-Flash-Next-Q4_K_S-00001-of-00003.gguf'
1.17.225.9xx W moe-cache-host: admission group=15..47 banks=3 bytes=~1.52-1.57GB reason=source_budget_exhausted cuda=none
1.18.537.850 I cmn init: llama threadpool init, n_threads = 16
1.20.188.219 I srv load_model: initializing, n_slots = 1, n_ctx_slot = 65536, kv_unified = 'false'
1.20.208.989 I srv llama_server: model loaded
1.20.208.993 I srv llama_server: listening on http://127.0.0.1:8080
2.08.430.966 I slot print_timing: id 0 | task 0 | prompt processing, n_tokens = 360, progress = 0.95, t = 5.78 s / 62.33 tokens per second
2.12.845.881 I slot print_timing: id 0 | task 0 | n_gen = 124, tg = 40.92 t/s, tg_3s = 41.25 t/s
2.21.877.055 I slot print_timing: id 0 | task 0 | n_gen = 515, tg = 42.70 t/s, tg_3s = 42.74 t/s
2.23.140.106 I srv operator (): operator (): cleaning up before exit...
```

The complete load-through-generation log is available.

## Current-branch retest status

This was reproduced on `907a73da9`. Current `moe-cache` contains later commit `16315a8fb` (`cuda: materialize pageable grouped MoE sources`), which may affect this path. The exact Windows 11 / RTX 4090 / 113 GB model workload has not yet been repeated on the current head.
